src/rdkafka_request.c

```diff
@@ -3328,6 +3328,284 @@ void rd_kafka_SaslAuthenticateRequest(rd_kafka_broker_t *rkb,
+}
+
+int rd_kafkap_leader_discovery_set_brokers(
+    rd_tmpabuf_t *tbuf,
+    rd_kafka_metadata_internal_t *mdi,
+    rd_kafkap_NodeEndpoints_t *NodeEndpoints) {
+        int i;
+        rd_kafka_metadata_t *md = &mdi->metadata;
+
+        size_t md_brokers_size =
+            NodeEndpoints->NodeEndpointCnt * sizeof(rd_kafka_metadata_broker_t);
+        size_t mdi_brokers_size = NodeEndpoints->NodeEndpointCnt *
+                                  sizeof(rd_kafka_metadata_broker_internal_t);
+
+        md->broker_cnt      = NodeEndpoints->NodeEndpointCnt;
+        md->brokers         = rd_tmpabuf_alloc(tbuf, md_brokers_size);
+        mdi->brokers_sorted = rd_tmpabuf_alloc(tbuf, md_brokers_size);
+        mdi->brokers        = rd_tmpabuf_alloc(tbuf, mdi_brokers_size);
+
+        if (!md->brokers || !mdi->brokers_sorted || !mdi->brokers)
+                return -1;
+
+        for (i = 0; i < NodeEndpoints->NodeEndpointCnt; i++) {
+                rd_kafkap_NodeEndpoint_t *NodeEndpoint =
+                    &NodeEndpoints->NodeEndpoints[i];
+                rd_kafka_metadata_broker_t *mdb           = &md->brokers[i];
+                rd_kafka_metadata_broker_internal_t *mdbi = &mdi->brokers[i];
+                mdb->id   = NodeEndpoint->NodeId;
+                mdb->host = NULL;
+                if (!RD_KAFKAP_STR_IS_NULL(&NodeEndpoint->Host)) {
+                        mdb->host = rd_tmpabuf_alloc(
+                            tbuf, RD_KAFKAP_STR_LEN(&NodeEndpoint->Host) + 1);
+                        rd_snprintf(mdb->host,
+                                    RD_KAFKAP_STR_LEN(&NodeEndpoint->Host) + 1,
+                                    "%.*s",
+                                    RD_KAFKAP_STR_PR(&NodeEndpoint->Host));
+                        if (!mdb->host)
+                                return -1;
+                }
+                mdb->port = NodeEndpoints->NodeEndpoints[i].Port;
+
+                /* Metadata internal fields */
+                mdbi->id      = mdb->id;
+                mdbi->rack_id = NULL;
+        }
+
+        qsort(mdi->brokers, md->broker_cnt, sizeof(mdi->brokers[0]),
+              rd_kafka_metadata_broker_internal_cmp);
+        memcpy(mdi->brokers_sorted, md->brokers,
+               sizeof(*mdi->brokers_sorted) * md->broker_cnt);
+        qsort(mdi->brokers_sorted, md->broker_cnt, sizeof(*mdi->brokers_sorted),
+              rd_kafka_metadata_broker_cmp);
+
+        return 0;
+}
+
+int rd_kafkap_leader_discovery_set_topic_cnt(rd_tmpabuf_t *tbuf,
+                                             rd_kafka_metadata_internal_t *mdi,
+                                             int topic_cnt) {
+
+        rd_kafka_metadata_t *md = &mdi->metadata;
+
+        md->topic_cnt = topic_cnt;
+        md->topics    = rd_tmpabuf_alloc(tbuf, sizeof(*md->topics) * topic_cnt);
+        mdi->topics = rd_tmpabuf_alloc(tbuf, sizeof(*mdi->topics) * topic_cnt);
+
+        if (!md->topics || !mdi->topics)
+                return -1;
```

Comment:
This can be removed as the `tmpabuf` is recreated with assert of failure equals to true

Reply:
Updated